akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

```diff
@@ -121,8 +119,12 @@ final class Flow[-In, +Out, +Mat](private[stream] override val module: Module)
    * Sink into the materialized value of the resulting Sink.
    */
   def toMat[Mat2, Mat3](sink: Graph[SinkShape[Out], Mat2])(combine: (Mat, Mat2) ⇒ Mat3): Sink[In, Mat3] = {
-    if (isIdentity) sink.asInstanceOf[Sink[In, Mat3]]
-    else {
+    if (isIdentity) {
+      if (sink.isInstanceOf[javadsl.Sink[In, Mat3]])
+        sink.asInstanceOf[javadsl.Sink[In, Mat3]].asScala
+      else
+        sink.asInstanceOf[Sink[In, Mat3]]
```

Comment:
We'll need calling combine here actually. I'll try to add some test case tomorrow.